src/group/libp2p_group_gossip_server.erl

```diff
@@ -109,41 +218,81 @@ handle_info(Msg, State) ->
 schedule_drop_timer(DropTimeOut) ->
     erlang:send_after(DropTimeOut, self(), drop_timeout).
 
--spec connections(Kind::atom() | all, #state{}) -> [{MAddr::string(), Pid::pid()}].
+-spec connections(libp2p_group_gossip:connection_kind() | all, #state{})
+                 -> [{MAddr::string(), Pid::pid()}].
 connections(Kind, #state{workers=Workers}) ->
-    lists:foldl(fun({_, _, undefined}, Acc) ->
+    lists:foldl(fun(#worker{target=undefined}, Acc) ->
                         Acc;
-                    ({_, Pid, MAddr}, Acc) when Kind == all ->
+                    (#worker{pid=Pid, target=MAddr}, Acc) when Kind == all ->
                         [{MAddr, Pid} | Acc];
-                    ({WorkerKind, Pid, MAddr}, Acc) when WorkerKind == Kind ->
+                    (#worker{kind=WorkerKind, pid=Pid, target=MAddr}, Acc) when WorkerKind == Kind ->
                         [{MAddr, Pid} | Acc];
                    (_, Acc) ->
                         Acc
                 end, [], Workers).
 
 assign_target(Kind, WorkerPid, TargetAddrs, State=#state{workers=Workers}) ->
-    SelectedAddr = case length(TargetAddrs) of
-                       0 -> undefined;
-                       _ -> mk_multiaddr(lists:nth(rand:uniform(length(TargetAddrs)), TargetAddrs))
-                   end,
-    libp2p_group_worker:assign_target(WorkerPid, SelectedAddr),
-    NewWorkers = lists:keyreplace(WorkerPid, 2, Workers, {Kind, WorkerPid, SelectedAddr}),
-    State#state{workers=NewWorkers}.
+    case length(TargetAddrs) of
```

Comment:
I thought you didn't like this refactor?

Reply:
meh.. 50/50, so I went with it.. it reads better than clearing the target